prov/verbs/src/verbs_info.c

```diff
@@ -1045,38 +1110,252 @@ static int fi_ibv_get_matching_info(uint32_t version, const char *dev_name,
+			           "failed to allocate src addr.\n");
+			return -FI_ENODATA;
+		}
+
+		if (flags & FI_SOURCE) {
+			if (service)
+				sscanf(service, "%" SCNu16, &src_addr->service);
+			VERBS_INFO(FI_LOG_CORE, "node '%s' service '%s' "
+				                "converted to <service=%d>\n",
+				   node, service, src_addr->service);
+		}
+	}
+
+	if (!dest_addr && node && !(flags & FI_SOURCE)) {
+		struct util_ns ns = (const struct util_ns){ 0 };
+		struct util_ns_attr ns_attr = {
+			.ns_port = fi_ibv_gl_data.dgram.name_server_port,
+			.name_len = sizeof(*dest_addr),
+			.service_len = sizeof(svc),
+			.service_cmp = fi_ibv_dgram_ns_service_cmp,
+			.is_service_wildcard = fi_ibv_dgram_ns_is_service_wildcard,
+		};
+
+		ret = ofi_ns_init(&ns_attr, &ns);
+		if (ret) {
+			VERBS_INFO(FI_LOG_CORE,
+				   "ofi_ns_init returns %d\n", ret);
+			ret = -FI_ENODATA;
+			goto fn1;
+		}
+
+		if (service)
+			svc = atoi(service);
+		svc0 = svc;
+		dest_addr = (struct ofi_ib_ud_ep_name *)
+			ofi_ns_resolve_name(&ns, node, &svc);
+		if (dest_addr) {
+			VERBS_INFO(FI_LOG_CORE,
+				   "'%s:%u' resolved to <gid <interface_id=%x, "
+				   "subnet_prefix=%x>, lid=%d>:%d\n",
+				   node, svc0, dest_addr->gid.global.interface_id,
+				   dest_addr->gid.global.subnet_prefix,
+				   dest_addr->lid, svc);
+		} else {
+			VERBS_INFO(FI_LOG_CORE,
+				   "failed to resolve '%s:%u'.\n", node, svc);
+			ret = -FI_ENODATA;
+			/* Here possible that `src_addr` isn't a NULL */
+			goto fn2;
+		}
+	}
+
+	for (iter_info = *info; iter_info; iter_info = iter_info->next) {
+		if (iter_info->ep_attr->type != FI_EP_DGRAM)
+			continue;
+		iter_info->addr_format = fmt;
+		if (src_addr) {
+			if (fmt == FI_ADDR_STR) {
+				size_t len = 0;
+				iter_info->src_addr =
+					fi_ibv_dgram_ep_name_to_string(src_addr,
+								       &len);
+				if (!iter_info->src_addr) {
+					ret = -FI_ENOMEM;
+					goto fn2;
+				}
+				iter_info->src_addrlen = len;
+			} else {
+				iter_info->src_addr = calloc(1, sizeof(*src_addr));
+				if (!iter_info->src_addr) {
+					ret = -FI_ENOMEM;
+					goto fn2;
+				}
+				memcpy(iter_info->src_addr, src_addr,
+				       sizeof(*src_addr));
+				iter_info->src_addrlen = sizeof(*src_addr);
+			}
+		}
+		if (dest_addr) {
+			if (fmt == FI_ADDR_STR) {
+				size_t len = 0;
+				iter_info->dest_addr =
+					fi_ibv_dgram_ep_name_to_string(dest_addr,
+								       &len);
+				if (!iter_info->dest_addr) {
+					ret = -FI_ENOMEM;
+					goto fn2;
+				}
+				iter_info->dest_addrlen = len;
+			} else {
+				iter_info->dest_addr = calloc(1, sizeof(*src_addr));
+				if (!iter_info->dest_addr) {
+					ret = -FI_ENOMEM;
+					goto fn2;
+				}
+				memcpy(iter_info->dest_addr, dest_addr, sizeof(*dest_addr));
+				iter_info->dest_addrlen = sizeof(*dest_addr);
+			}
+		}
+	}
+
+	/* `fi_info::src_addr` and `fi_info::dest_addr` is freed
+	 * in the `fi_freeinfo` function */
+fn2:
+	if (src_addr)
+		free(src_addr);
+	if (dest_addr)
+		free(dest_addr);
+fn1:
+	return ret;
+}
```

Comment:
nit: This is a really long function.  Is there a way to break it up?

Reply:
Yes, there is some way to break it up.
Will upload new version of the function.